File: src/models/model_info.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ModelType {
    Chat,
    Completion,
    Embedding,
    ImageGeneration,
    Audio,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelInfo {
    pub id: String,
    pub name: String,
    pub context_length: u32,
    pub max_tokens: u32,
    pub vision: bool,
    pub function_call: bool,
    pub reasoning: bool,
    #[serde(rename = "type")]
    pub model_type: ModelType,
}

impl ModelInfo {
    pub fn new(
        id: String,
        name: String,
        context_length: u32,
        max_tokens: u32,
        vision: bool,
        function_call: bool,
        reasoning: bool,
        model_type: ModelType,
    ) -> Self {
        Self {
            id,
            name,
            context_length,
            max_tokens,
            vision,
            function_call,
            reasoning,
            model_type,
        }
    }
}

File: src/models/model_info.ts
export enum ModelType {
  Chat = "chat",
  Completion = "completion",
  Embedding = "embedding",
  ImageGeneration = "imagegeneration",
  Audio = "audio",
}

export interface ModelModalities {
  input?: string[];
  output?: string[];
  [key: string]: string[] | undefined;
}

export interface ModelCost {
  input?: number;
  output?: number;
  cacheRead?: number;
  cacheWrite?: number;
  [key: string]: number | undefined;
}

export interface ModelLimit {
  context?: number;
  output?: number;
  requestsPerMinute?: number;
  requestsPerDay?: number;
  [key: string]: number | undefined;
}

export interface ModelInfo {
  id: string;
  name: string;
  contextLength: number;
  maxTokens: number;
  vision: boolean;
  functionCall: boolean;
  reasoning: boolean;
  type: ModelType;
  attachment?: boolean;
  temperature?: boolean;
  toolCall?: boolean;
  knowledge?: string;
  releaseDate?: string;
  lastUpdated?: string;
  openWeights?: boolean;
  experimental?: boolean;
  modalities?: ModelModalities;
  cost?: ModelCost;
  limit?: ModelLimit;
  metadata?: Record<string, unknown>;
}

export function createModelInfo(
  id: string,
  name: string,
  contextLength: number,
  maxTokens: number,
  vision: boolean,
  functionCall: boolean,
  reasoning: boolean,
  modelType: ModelType,
  overrides: Partial<Omit<ModelInfo, 'id' | 'name' | 'contextLength' | 'maxTokens' | 'vision' | 'functionCall' | 'reasoning' | 'type'>> = {},
): ModelInfo {
  return {
    id,
    name,
    contextLength,
    maxTokens,
    vision,
    functionCall,
    reasoning,
    type: modelType,
    ...overrides,
  };
}
